include/deal.II/base/index_set.h

```diff
@@ -229,6 +229,12 @@ class IndexSet
   void compress () const;
 
   /**
+   * Reorder the ranges inside IndexSet so that the largest range is the first
+   * one.
+   */
+  void reorder_ranges();
```

Comment:
After thinking about this some more, how about this idea:

* IndexSet internally stores its ranges in ascending sorted order; this is as it currently already does, and makes many operations efficient
* It also stores a pointer to one of the ranges. `compress()` at the very ends sets this pointer to the largest range it encounters, but for correctness it really only has to point *somewhere*. This makes sure that we don't have to change the entire code base and break one of the current invariants (that the set of ranges is sorted).
* For time-critical operations, the implementing function would check the range pointed to first, before going through the remainder (or in fact all) of the ranges.